### PR TITLE
Add supporting neo4j v4 syntax

### PR DESCRIPTION
**Pull request by dzianis-hrynko**

### CODE REVIEW

File: src/Persister/EntityPersister.php [hunk 1]
```diff
@@ -143,17 +93,50 @@ public function refresh(int $id, object $entity)
 
     public function getDetachDeleteQuery($object): Statement
     {
-        $query = 'MATCH (n) WHERE id(n) = {id} DETACH DELETE n';
+        $query = sprintf("MATCH (n) WHERE id(n) = $this->paramStyle DETACH DELETE n", 'id');
```
Comment (gep) [inline src/Persister/EntityPersister.php]: I don't understand what is the point of sprintf here?

File: src/Persister/EntityPersister.php [hunk 2]
```diff
@@ -143,17 +93,50 @@ public function refresh(int $id, object $entity)
...
         $id = $this->classMetadata->getIdValue($object);
 
         return Statement::create($query, ['id' => $id]);
     }
 
     public function getDeleteQuery($object): Statement
     {
-        $query = 'MATCH (n) WHERE id(n) = {id} DELETE n';
+        $query = sprintf("MATCH (n) WHERE id(n) = $this->paramStyle DELETE n", 'id');
```
Comment (gep) [inline src/Persister/EntityPersister.php]: Do we need `sprintf` here?

File: src/Persister/EntityPersister.php [hunk 3]
```diff
@@ -143,17 +93,50 @@ public function refresh(int $id, object $entity)
...
         $id = $this->classMetadata->getIdValue($object);
 
         return Statement::create($query, ['id' => $id]);
     }
+
+    private function getBaseQueryProperties($object)
```
Comment (gep) [inline src/Persister/EntityPersister.php]: Can you please add more typing here?

File: src/Persister/EntityPersister.php [hunk 4]
```diff
@@ -143,17 +93,50 @@ public function refresh(int $id, object $entity)
...
+            $fieldId = $this->classMetadata->getClassName() . $field;
+            $fieldKey = $field;
+
+            if ($meta->getPropertyAnnotationMetadata()->hasCustomKey()) {
+                $fieldKey = $meta->getPropertyAnnotationMetadata()->getKey();
+            }
+
+            if ($meta->hasConverter()) {
+                $converter = Converter::getConverter($meta->getConverterType(), $fieldId);
+                $v = $converter->toDatabaseValue($meta->getValue($object), $meta->getConverterOptions());
```
Comment (gep) [inline src/Persister/EntityPersister.php]: This:

```php
$v = $converter->toDatabaseValue($meta->getValue($object), $meta->getConverterOptions());
                $propertyValues[$fieldKey] = $v;
```

May be written as

```php
$propertyValues[$fieldKey] = $converter->toDatabaseValue($meta->getValue($object), $meta->getConverterOptions());
```

File: src/Persisters/BasicEntityPersister.php [hunk 5]
```diff
@@ -121,11 +121,11 @@ public function getMatchCypher(
 
         $filter_cursor = 0;
         $params = [];
-
+        $paramStyle = $this->entityManager->isV4() === true ? '$%s' : '{%s}';
         foreach ($criteria as $key => $criterion) {
             $key = (string) $key;
             $clause = $filter_cursor === 0 ? 'WHERE' : 'AND';
-            $cypher .= sprintf('%s %s.%s = {%s} ', $clause, $identifier, $key, $key);
+            $cypher .= sprintf("%s %s.%s = $paramStyle ", $clause, $identifier, $key, $key);
```
Comment (gep) [inline src/Persisters/BasicEntityPersister.php]: Please, add it to sprintf as a string

File: src/Query.php [hunk 6]
```diff
@@ -116,6 +118,9 @@ public function getOneOrNullResult(): ?array
     public function execute(): array
     {
         $stmt = $this->getCql();
+        if ($this->entityManager->isV4() === true) {
```
Comment (gep) [inline src/Query.php]: Why do we need this?

File: src/Query.php [hunk 7]
```diff
@@ -268,4 +273,19 @@ private function hydrateMapValue(int|string $key, mixed $value): array|null|int|
         }
         return $row;
     }
+
+    private function queryVersionConverter(string $query): string
+    {
+        preg_match_all(self::PARAM_V3_REG_EXP, $query, $matches);
```
Comment (gep) [inline src/Query.php]: can we use preg_replace instead?

File: src/Persister/EntityPersister.php (hunk 2 above)
Comment (dzianis-hrynko) [reply src/Persister/EntityPersister.php]: yes, for switching between verisons

File: src/Query.php (hunk 7 above)
Comment (dzianis-hrynko) [reply src/Query.php]: Usually, $query has more than one match, I need to transform them all

File: src/Persister/EntityPersister.php (hunk 1 above)
Comment (dzianis-hrynko) [reply src/Persister/EntityPersister.php]: for switching between verisons

File: src/Query.php (hunk 6 above)
Comment (dzianis-hrynko) [reply src/Query.php]: for switching between verisons

File: src/Query.php (hunk 7 above)
Comment (gep) [reply src/Query.php]: So, preg_replace can replace everything without any additional loops and string manipulaitons

File: src/Query.php [hunk 8]
```diff
@@ -116,6 +118,9 @@ public function getOneOrNullResult(): ?array
     public function execute(): array
     {
         $stmt = $this->getCql();
+        if ($this->entityManager->isV4()) {
+            $stmt = $this->queryVersionConverter($stmt);
+        }
```
Comment (gep) [inline src/Query.php]: This:

```php
$stmt = $this->getCql();
        if ($this->entityManager->isV4()) {
            $stmt = $this->queryVersionConverter($stmt);
        }
```

may be written as 

```php
$stmt = $this->entityManager->isV4())
  ? $this->queryVersionConverter($stmt)
  : $this->getCql();
```

**Pull request merged**